group-coordinator/src/main/java/org/apache/kafka/coordinator/group/generic/GenericGroup.java

```diff
@@ -898,6 +904,53 @@ public void createGroupTombstoneRecords(List<Record> records) {
         records.add(RecordHelpers.newGroupMetadataTombstoneRecord(groupId()));
     }
 
+    @Override
+    public boolean isEligibleForDeletion() {
+        return isInState(EMPTY) && generationId > 0;
```

Comment:
Are you sure about the `generationId > 0` condition here? I cannot find it. There may be some confusion with [this one](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala#L888). I think that this optimization does not apply to us because we need a record to remove the group from the timeline data structure.

Reply:
that's what i'm confused with; in the existing coordinator, we remove from the cache but do not append tombstones. doesn't this mean the empty group will be populated in the cache when the coordinator restarts?

Reply:
Probably not because we don't create a record for those groups either. In our case, the issue is that the group won't be remove from the cache if we don't have a tombstone for it because we update the cache when we replay records.

Reply:
ah got it. makes sense